src/main/java/Parser.java

```diff
@@ -0,0 +1,102 @@
+import java.util.Scanner;
+
+/**
+ * Handles user commands.
+ */
+public class Parser {
+    private Ui ui;
+    private TaskList taskList;
+    private Storage storage;
+
+    /**
+     * Constructor for Parser.
+     * @param ui ui instance.
+     * @param taskList tasklist instance.
+     * @param storage storage instance.
+     */
+    public Parser(Ui ui, TaskList taskList, Storage storage) {
+        this.ui = ui;
+        this.taskList = taskList;
+        this.storage = storage;
+    }
+
+    /**
+     * Runs actions invoked by user input.
+     * @throws DukeException  If invalid command such as empty command.
+     */
+
+    public String parse(String lines) {
```

Comment:
Ideally, the Parser class and the parse function should only serve to parse the user input to recognise different commands, instead of modifying the tasklist directly! :) But good organisation and structure overall 👍